crates/red_knot_workspace/src/project/metadata.rs

```diff
@@ -67,21 +87,58 @@ impl ProjectMetadata {
 
         let mut closest_project: Option<ProjectMetadata> = None;
 
-        for ancestor in path.ancestors() {
-            let pyproject_path = ancestor.join("pyproject.toml");
-            if let Ok(pyproject_str) = system.read_to_string(&pyproject_path) {
-                let pyproject = PyProject::from_str(&pyproject_str).map_err(|error| {
-                    ProjectDiscoveryError::InvalidPyProject {
-                        path: pyproject_path,
-                        source: Box::new(error),
+        for project_root in path.ancestors() {
+            let pyproject_path = project_root.join("pyproject.toml");
+
+            let pyproject = if let Ok(pyproject_str) = system.read_to_string(&pyproject_path) {
+                match PyProject::from_toml_str(&pyproject_str) {
+                    Ok(pyproject) => Some(pyproject),
+                    Err(error) => {
+                        return Err(ProjectDiscoveryError::InvalidPyProject {
+                            path: pyproject_path,
+                            source: Box::new(error),
+                        })
+                    }
+                }
```

Comment:
`map_err` didn't work here (and below). The borrow checker gets mad at me for consuming `pyproject_path` in the closure and using it further down.